### PR TITLE
[mlir][gpu] Pack and unpack to enable f16 and int8 warp reduce.

### PR DESCRIPTION
**Pull request by raikonenfnu**

### CODE REVIEW

**Comment by raikonenfnu** [comment]
@ThomasRaoux here is a draft PR, would like your comments and suggestions, for now on my todo is:

1. Pack everything to i32 rather than f32 when float(f16 -> f32) and i32 when int(i.e i8 -> i32)
2. Write extra multi warp test for the low-p(f16/i8) to test that the broadcast to reduce num of shuffles still works.

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 1]
```diff
@@ -40,28 +42,77 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
   return builder.create<memref::AllocOp>(loc, memrefType);
 }
 
+/// Packs vector of lower precision into a single 32-bit width element.
+/// (i.e <2xf16> -> f32 and <4xi8> -> i32)
+static Value packVectorToSupportedWidth(Location loc, OpBuilder &builder,
+                                        Value input) {
+  auto vecType = input.getType().cast<VectorType>();
+  assert(vecType && "This packing fn only works if input is a vector.");
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:50]: nit: `cast` already asserts that the cast is correct and never return null

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 2]
```diff
@@ -40,28 +42,77 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
...
+  Type elementType = vecType.getElementType();
+  int64_t vecSize = vecType.getShape()[0];
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:52]: nit: `vecType.getDimSize(0)`

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 3]
```diff
@@ -40,28 +42,77 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
...
+  unsigned bitWidth = elementType.getIntOrFloatBitWidth();
+  // Check that numel * bitwidth = 32.
+  assert(vecSize * bitWidth == kShuffleBitWidth &&
+         "vector needs to packable into 32-bitwidth.");
+  assert(elementType.isIntOrFloat() && "Only int and float supported.");
+  Type supportedType = builder.getF32Type();
+  if (elementType.dyn_cast<IntegerType>()) {
+    supportedType = builder.getI32Type();
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:60]: as discussed I would use i32 all the time

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 4]
```diff
@@ -40,28 +42,77 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
...
     Value shuffled = builder
-                         .create<gpu::ShuffleOp>(loc, laneVal, i,
+                         .create<gpu::ShuffleOp>(loc, shuffleInput, i,
                                                  /*width=*/warpSize,
                                                  /*mode=*/gpu::ShuffleMode::XOR)
                          .getShuffleResult();
+    if (unpackedType) {
+      Value vectorShuffled = promoteElementToVector(loc, builder, shuffled);
+      shuffled =
+          builder.create<vector::BitCastOp>(loc, unpackedType, vectorShuffled);
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:98]: nit: I would make it a unpackVector symmetrical to `packVectorToSupportedWidth`

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 5]
```diff
@@ -151,12 +274,17 @@ static Value groupReduction(Location loc, OpBuilder &builder, Value input,
       Attribute identityAttr =
           getCombiningKindIdentity(builder, kind, loadVal.getType());
       assert(identityAttr && "Unknown identity value for the reduction");
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:276]: dead code?

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 6]
```diff
@@ -151,12 +274,17 @@ static Value groupReduction(Location loc, OpBuilder &builder, Value input,
...
-      Value identity = builder.create<arith::ConstantOp>(loc, identityAttr);
+      Value identity =
+          getCombiningIdentityValue(loc, builder, kind, loadVal.getType());
       loadVal = builder.create<arith::SelectOp>(loc, useIdentityElement,
                                                 identity, loadVal);
     }
     laneVal = warpReduction(loc, builder, loadVal, kind, warpSize, numWarp);
   }
+  // Handles cases for sub-32bit precision where output is still in vector form.
+  if (laneVal.getType().dyn_cast<VectorType>()) {
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:285]: nit: should be `laneVal.getType().isa<VectorType>()`

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 7]
```diff
@@ -105,6 +156,78 @@ static Attribute getCombiningKindIdentity(OpBuilder &builder,
...
+      Value reductionSlice =
+          builder.create<vector::ReductionOp>(loc, kind, laneValSlice);
+      SmallVector<int64_t> perLaneUnrollId = {i};
+      perLaneReduction = builder.create<vector::InsertOp>(
+          loc, reductionSlice, perLaneReduction, perLaneUnrollId);
+    }
+  } else {
+    // In cases where vecSize < unrollCount, we would pad the vector
+    // with identity elements until it's total bit size is 32.
+    Attribute identityAttr =
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:200]: is there a test for this case? Can we add one?

**Review by ThomasRaoux — COMMENTED** [review]
Looks great, added a bunch of style related comments. Can you add an execution test for the new types here:
https://github.com/iree-org/iree/blob/main/tests/e2e/regression/large_reduction.mlir

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 2 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 4 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 5 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: yeap, good catch :)

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 6 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 7 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:209]: digging up for the dispatches by looking through the dispatches of SD 😅

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 8]
```diff
@@ -20,6 +20,8 @@
 
 #define DEBUG_TYPE "iree-codegen-reduction-distribution"
 
+constexpr unsigned kShuffleBitWidth = 32;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:23]: nit: make it static to not pollute the global namespace.

File: compiler/src/iree/compiler/Codegen/Common/test/warp_reduction.mlir [hunk 9]
```diff
@@ -78,6 +78,204 @@ hal.executable private @simple_reduce  {
...
+      %cst_0 = arith.constant 0.000000e+00 : f16
+      %cst_1 = arith.constant dense<3.840000e+02> : vector<1xf16>
+      %c32 = arith.constant 32 : index
+      %c384 = arith.constant 384 : index
+      %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : memref<128x384xf16>
+      %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : memref<128xf16>
+      %workgroup_id_x = hal.interface.workgroup.id[0] : index
+      %2 = gpu.thread_id  x
+      %3 = affine.apply affine_map<()[s0, s1] -> (s1 * 2 + s0 floordiv 32)>()[%2, %workgroup_id_x]
+      %4 = scf.for %arg0 = %c0 to %c384 step %c32 iter_args(%arg1 = %cst) -> (vector<1xf16>) {
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/test/warp_reduction.mlir:106]: optional: you don't really need the loop here, it is better to have reduced tests.

**Review by ThomasRaoux — COMMENTED** [review]
LGTM once the last nit are addressed

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp [hunk 10]
```diff
@@ -40,28 +42,86 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
...
+/// Packs vector of lower precision into a single 32-bit width element.
+/// (i.e <2xf16> -> i32 and <4xi8> -> i32)
+static Value packVectorToSupportedWidth(Location loc, OpBuilder &builder,
+                                        Value input) {
+  auto vecType = input.getType().cast<VectorType>();
+  Type elementType = vecType.getElementType();
+  assert(vecType.getDimSize(0) * elementType.getIntOrFloatBitWidth() ==
+             kShuffleBitWidth &&
+         "vecSize * vecBitWidth needs to packable into 32-bitwidth.");
+  assert(elementType.isIntOrFloat() &&
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:64]: @ThomasRaoux Any thoughts on how to handle asserts with very long variable and getter names. For example, currently this is hitting vecType and elementType is not being used on certain IREE builders/tests. On the other hand if I put in the full getter (i.e input.getType().cast<VectorType>() getElementType(). getIntOrFloatBitWidth() for elementType's bitWidth), feels a bit long and unreadable 😅. Thanks in advance :)

File: compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp (hunk 7 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp:213]: done, simple_half_reduce_small is supposed to represent it :)

File: compiler/src/iree/compiler/Codegen/Common/test/warp_reduction.mlir (hunk 9 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/test/warp_reduction.mlir]: done :)

**Review by ThomasRaoux — APPROVED** [review]

**Pull request merged**